infowav.1
.TH infowav 1 "October 1999" Linux "Linux Audio Manual"
.SH NAME
infowav \- WAV File Playback and Analysis Tool
.SH SYNOPSIS
.B infowav [-v][-i][-l][-p <dsp-dev>] <wav-file>

.SH DESCRIPTION
\fBInfowav\fP is a command-line tool for analyzing audio files in the
Microsoft WAV format.  Both basic and extended attribute information can
be viewed, as well as playback of the actual audio.

.SH OPTIONS
.TP
.B \-v
Print a version identifier and then exit.
.TP
.B \-i
Print basic audio format information, such as sample rate, sample size, etc.
.TP
.B \-l
Print extended list attribute information.
.TP
.B \-p <dsp-dev>
Play the audio, using \fI<dsp-dev>\fP as the playback device.  If not 
specified, \fI/dev/dsp\fP will be used by default.
.SH NOTES
The meaning of fields in the extended list attributes tend to be very
vendor-dependent.
.SH TODO
Currently, \fBinfowav\fP is capable of fully deciphering only files in 
Linear PCM format.  MP3 support would be a Good Thing.
.SH AUTHOR
Fred Gleason <fredg@wava.com>


History
                    History file for the WAVLIB Project

-------------------------------------------------------------------------------
11/20/1999
Start of history file

Current released version is 2.2.2

Added the aproc(1) command, a command-line audio processor for wav files.
Currently, it only supports the 'normalize' function (-N), but is intended
as a catch-all for adding additional processing features as needed through
addition of additional command flags.

Added a man page for the aproc(1) command.

Fixed a bug in OpenWav() that was causing the file pointer to not be
pointing at the start of the wave data at function return with certain 
combinations of chunks. 

Fixed a typo in the ameter(1) man page.
-------------------------------------------------------------------------------

12/1/1999 -- v 2.3.1
Added a conditional definition of PI (definition is missing from glibc).
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.0

Cleaned up code in ameter(1) so that levels are now displayed in true 
logarithmic dB.

Added a '-l' option to ameter(1) to allow the user to set the meter reference
level

Added color to ameter(1), with a green and red segment.  Added a '-t' option
to allow the user to define where the border should be.

Updated the ameter(1) man page.

Fixed a bug in infowav(1) that was causing it to corrupt wav files when invoked
with the '-p' while omitting the audio-device name that customarily follows it.

Fixed a bug in infowav(1) that was causing extraneous output to be printed.

Added a '-t' option to recwav(1) to allow the user to specify a record length.

Updated the recwav(1) man page.

Added a TailTrim() API call to remove audio below a given set threshold level
from the end of an existing wav file.
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.1

Restored the RecordWavFile() API call, which somehow got dropped out of
v2.4.0.
-------------------------------------------------------------------------------

5/23/2000 -- v 2.4.2

Fixed a fencepost bug in TailTrim() that was causing false positive 
threshold detects on certain wav files.
-------------------------------------------------------------------------------

infowav.c
/* infowav.c
 *
 *  A general-purpose utility for analyzing files in the the Microsoft
 *  'wav' format.
 *
 *   (C) Copyright 2002-2005 Fred Gleason <fredg@salemradiolabs.com>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation; either version 2 of
 *   the License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public
 *   License along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

#include "wavlib.h"
#define USAGE "infowav [-v][-i][-l][-p <dsp-dev>] <wav-file>"


int main(int argc,char *argv[])
{
  int hFilename;
  struct wavHeader wavHeader;
  struct wavChunk wavChunk;
  struct wavList wavList;
  int i;
  int dInfo=0,dList=0,dPlay=0,dOptionFound;
  char sDspdev[256];
  char sFormat[256];


  /*
   * Parse Options
   */
  if(argc<2) {
    printf(USAGE);
    printf("\n");
    exit(1);
  }
  if(argc==2) {
    if(strcmp(argv[1],"-v")==0) {   /* Version */
      printf("infowav v");
      printf(VERNUM);
      printf("\n");
      exit(0);
    }
    else {
      printf("infowav: invalid argument\n");
      exit(1);
    }
  }
  for(i=1;i<argc-1;i++) {
    dOptionFound=0;
    if(strcmp(argv[i],"-i")==0) {    /* Display General Info */
      dInfo=1;
      dOptionFound=1;
    }
    if(strcmp(argv[i],"-l")==0) {    /* Display Extended List Data */
      dList=1;
      dOptionFound=1;
    }
    if(strcmp(argv[i],"-p")==0) {    /* Play on Specified Device */
      dPlay=1;
      strcpy(sDspdev,argv[++i]);
      if(i==argc-1) {
	printf("infowav: missing wav file argument\n");
	exit(1);
      }
      dOptionFound=1;
    }
    if(dOptionFound==0) {
      printf("infowav: invalid option\n");
      exit(1);
    }
  }

  /*
   * Initialize
   */
  hFilename=OpenWav(argv[argc-1],&wavHeader);
  if(hFilename<0) {
    perror("infowav");
    exit(1);
  }

  /*
   * Display Info
   */
  if(dInfo==1) {
  printf("\n");
  printf("         Wave File: %s\n",argv[argc-1]);
    switch(wavHeader.wFormatTag) {
    case WAVE_FORMAT_PCM:
      strcpy(sFormat,"Linear PCM");
      break;
    case IBM_FORMAT_MULAW:
      strcpy(sFormat,"IBM Mu-Law");
      break;
    case IBM_FORMAT_ALAW:
      strcpy(sFormat,"IBM A-Law");
      break;
    case IBM_FORMAT_ADPCM:
      strcpy(sFormat,"IBM Adaptive-Differential PCM");
      break;
    default:
      strcpy(sFormat,"Unknown");
    }
    printf("     Sample Format: %s (0x%04X)\n",sFormat,(int)wavHeader.wFormatTag);
    printf("          Channels: %d\n",(int)wavHeader.wChannels);
    printf("  Samples/sec/chan: %u\n",wavHeader.dwSamplesPerSec);
    printf("       Bits/sample: %d\n",(int)wavHeader.wBitsPerSample);
    printf("         Bytes/sec: %u\n",wavHeader.dwAvgBytesPerSec);
    printf("   Block Alignment: %d\n",(int)wavHeader.wBlockAlign);
    printf("     Total Samples: %u\n",wavHeader.dwFileSize);
    printf("      Audio Length: %u sec\n",(unsigned)wavHeader.tWavLength);
    close(hFilename);
    printf("       *** CHUNK DATA ***\n");
    printf("Name   --Offset--    ---Size---\n");
    hFilename=open(argv[argc-1],O_RDONLY);
    if(hFilename<0) {
      perror("infowav");
      exit(1);
    }
    lseek(hFilename,12,SEEK_SET);
    while(GetNextChunk(hFilename,&wavChunk)==0) {
      printf("%4s   %10u    %10u\n",wavChunk.sName,(unsigned)wavChunk.oOffset,
	     (unsigned)wavChunk.oSize);
    }
  }

  if(dList==1) {  /* Display extended list data */
    i=GetListChunk(hFilename,&wavList);
    if(i<0) {
      printf("  No extended list attributes found.\n");
    }
    else {
      printf("  *** EXTENDED LIST ATTRIBUTES ***\n");
      if(i==0) {
	printf("ICRD: %s\n",wavList.sIcrd);
      }
      else {
	printf("ICRD: --NOT PRESENT--\n");
      }
      printf("IART: %s\n",wavList.sIart);
      printf("ICMT: %s\n",wavList.sIcmt);
      printf("ICOP: %s\n",wavList.sIcop);
      printf("IENG: %s\n",wavList.sIeng);
      printf("IGNR: %s\n",wavList.sIgnr);
      printf("IKEY: %s\n",wavList.sIkey);
      printf("IMED: %s\n",wavList.sImed);
      printf("INAM: %s\n",wavList.sInam);
      printf("ISFT: %s\n",wavList.sIsft);
      printf("ISRC: %s\n",wavList.sIsrc);
      printf("ITCH: %s\n",wavList.sItch);
      printf("ISBJ: %s\n",wavList.sIsbj);
      printf("ISRF: %s\n",wavList.sIsrf);
    }
  }

  close(hFilename);

  if(dPlay==1) {     /* Play the wavfile */
    printf("\n");
    printf("Playing: %s\n",argv[argc-1]);
    if(PlayWavFile(argv[argc-1],sDspdev,0)<0) {
      perror("infowav");
      exit(1);
    }
  }


  exit(0);
}
